src/rendering/SceneManager.cc

```diff
@@ -1202,6 +1203,119 @@ rendering::VisualPtr SceneManager::CreateInertiaVisual(Entity _id,
   return inertiaVis;
 }
 
+/////////////////////////////////////////////////
+rendering::VisualPtr SceneManager::CreateJointVisual(
+    Entity _id, const sdf::Joint &_joint,
+    Entity _childId, Entity _parentId)
+{
+  if (!this->dataPtr->scene)
+  {
+    return rendering::VisualPtr();
+  }
+
+  if (this->dataPtr->visuals.find(_id) != this->dataPtr->visuals.end())
+  {
+    ignerr << "Entity with Id: [" << _id << "] already exists in the scene"
+           << std::endl;
+    return rendering::VisualPtr();
+  }
+
+  rendering::VisualPtr parent;
+  if (_childId != this->dataPtr->worldId)
+  {
+    auto it = this->dataPtr->visuals.find(_childId);
+    if (it == this->dataPtr->visuals.end())
+    {
+      // It is possible to get here if the model entity is created then
+      // removed in between render updates.
+      return rendering::VisualPtr();
+    }
+    parent = it->second;
+  }
+
+  // Name.
+  std::string name = _joint.Name().empty() ? std::to_string(_id) :
+    _joint.Name();
+  if (parent)
+  {
+    name = parent->Name() +  "::" + name;
+  }
+
+  rendering::JointVisualPtr jointVisual =
+    this->dataPtr->scene->CreateJointVisual(name);
+
+  switch (_joint.Type())
+  {
+    case sdf::JointType::REVOLUTE:
+      jointVisual->SetType(rendering::JointVisualType::JVT_REVOLUTE);
+      break;
+    case sdf::JointType::REVOLUTE2:
+      jointVisual->SetType(rendering::JointVisualType::JVT_REVOLUTE2);
+      break;
+    case sdf::JointType::PRISMATIC:
+      jointVisual->SetType(rendering::JointVisualType::JVT_PRISMATIC);
+      break;
+    case sdf::JointType::UNIVERSAL:
+      jointVisual->SetType(rendering::JointVisualType::JVT_UNIVERSAL);
+      break;
+    case sdf::JointType::BALL:
+      jointVisual->SetType(rendering::JointVisualType::JVT_BALL);
+      break;
+    case sdf::JointType::SCREW:
+      jointVisual->SetType(rendering::JointVisualType::JVT_SCREW);
+      break;
+    case sdf::JointType::GEARBOX:
+      jointVisual->SetType(rendering::JointVisualType::JVT_GEARBOX);
+      break;
+    case sdf::JointType::FIXED:
+      jointVisual->SetType(rendering::JointVisualType::JVT_FIXED);
+      break;
+    default:
+      jointVisual->SetType(rendering::JointVisualType::JVT_NONE);
+      break;
+  }
+
+  if (parent)
+  {
+    jointVisual->RemoveParent();
+    parent->AddChild(jointVisual);
+  }
+
+  if (_joint.Axis(1))
+  {
+    auto axis1 = _joint.Axis(0)->Xyz();
+    auto axis2 = _joint.Axis(1)->Xyz();
+    auto axis1UseParentFrame =
+        (_joint.Axis(0)->XyzExpressedIn() == "__model__")? true : false;
+    auto axis2UseParentFrame =
+        (_joint.Axis(1)->XyzExpressedIn() == "__model__")? true : false;
+
+    jointVisual->SetAxis(axis2, axis2UseParentFrame);
+
+    if (this->dataPtr->visuals.find(_parentId) != this->dataPtr->visuals.end())
+    {
+      auto parentName = this->dataPtr->visuals[_parentId]->Name();
```

Comment:
nitpick: store the find result to avoid a second lookup:

```cpp
    auto it = this->dataPtr->visuals.find(_parentId);
    if (it != this->dataPtr->visuals.end())
    {
      auto parentName = it->second->Name();
```

Reply:
[`de37a0a`](https://github.com/ignitionrobotics/ign-gazebo/pull/961/commits/de37a0adce1fb2d6d1e2f171a10dcd5d4b9c2971)